rasa/shared/core/domain.py

```diff
@@ -1476,9 +1476,11 @@ def _transform_intents_for_file(self) -> List[Union[Text, Dict[Text, Any]]]:
                 if rasa.shared.core.constants.ENTITY_LABEL_SEPARATOR not in entity
             )
             ignore_entities = set(self.entities) - use_entities
-            if len(use_entities) == len(self.entities):
+            use_length = len(use_entities)
+            self_length = len(self.entities)
+            if use_length == self_length:
                 intent_props[USE_ENTITIES_KEY] = True
-            elif len(use_entities) <= len(self.entities) / 2:
+            elif use_length > self_length or use_length <= self_length / 2:
```

Comment:
I might be wrong here but I think what we need to do here is to merge entities from all domain files to be able to use `intent_props[IGNORE_ENTITIES_KEY]`. Otherwise whenever we fall into the `else` 2 lines below, we'll be using a subset of ignored entities and it will be wrong, ie. other entities defined in the other domain files will be missing and hence not in ignore_entities and they will not be ignored. Does that make sense? Can we add a test where second domain adds 5 entities, uses 4 of them for an intent, and hence it will fall into `else` below but it will still have a separate intent defined in the first domain in ignored list? This is hard to explain but let me give examples:

Domain1:
```
intents:
  - greet:
      use_entities: [city]
 
entities:
  - city
```
Domain2:
```
intents:
  - goodbye:
      use_entities: [time, price, gender, date]
 
entities: [time, price, gender, date, not_used_entity]
```
The test should check that ignored entities for `goodbye` intent includes `city` even though it is not used in domain2 (I think it will correctly include `not_used_entity`, but would be good to assert that's there too).

Reply:
I'm actually not sure about this method here, as it seems to be used only for display or writing to a file. It looks as though `_transform_intent_properties_for_internal_use` is the method which transforms user domain files to internal format. For display/file output, we may not need to know about other domain files, but while using intents internally we do (because config files are merged and used all together).

Reply:
@tayfun You are perfectly right. We must not make a Domain object before merging all dicts. See issue #10137 comments.

Reply:
> The test should check that ignored entities for goodbye intent includes city even though it is not used in domain2 (I think it will correctly include not_used_entity, but would be good to assert that's there too).

@tayfun @raoulvm Does that mean that ignored entities for `greet` should include `time`, `price`, `gender`, `date`, and `not_used_entity`?

And what if `greet` does not specify any entities:
```
intents:
  - greet:
 
entities:
  - city
 ```
Usually this would mean all defined entities are used. So would it use _all_ entities (`city`, `greet` should include `time`, `price`, `gender`, `date`, and `not_used_entity) or just the entity defined within the same file, `city`?

Reply:
Hi @carlad from my point of view you are right in both cases.

Reply:
thanks @raoulvm so the behaviour should be that _any_ `entities`, no matter which domain file they are specified in, should be assigned to _any_ `intent`, unless explicitly `use`d or `ignore`d. Further, any explicit use automatically ignores remaining entities, and vice versa, where any explicit ignore, automatically uses remaining entities.

Reply:
Okay - so I think I know what's happening that caused the original issue.

In a domain file, when an intent is set, with a `use_entity` configured, and that same entity is the _only_ entity specified in the file, eg:
```
intents:
  - greet:
      use_entities: [city]
 
entities:
  - city
```
This is interpreted as `'intents': [{'greet': {'use_entities': True}}]` because it is using all the entities as specified at that point. 

When the subsequent domain files are merged, the value of `greet` stays the same, i.e. ` {'use_entities': True}}` and it uses all the entities.

I want to look more into the existing `Domain.merge` method to see if we can implement a solution to this there.

Reply:
No you can't because for that you need to create a Domain object from the previous file, which then looses the information on which entities to use. "used_entities" is a positive only list. In @tayfun example above `greet` uses one entity out of one. That is, concerning the Domain.to_dict(), all of them. But given that the to-be-merged file also contains entities, that information "use exactly that one entity" is lost. Flipside, an intent using all entities would get the same "one of one" entity attribute in "used_entities". So both would have either all or only `city` entity.